src/adapter/src/client.rs

```diff
@@ -363,6 +368,44 @@ impl SessionClient {
         self.session = Some(res.session);
         res.result
     }
+
+    pub fn add_idle_in_transaction_session_timeout(&mut self) {
+        let session = self.session();
+        let timeout_dur = session.vars().idle_in_transaction_session_timeout();
+        if !timeout_dur.is_zero() {
+            let timeout_dur = timeout_dur.clone();
+            if let Some(txn) = session.transaction().inner() {
+                let txn_id = txn.id.clone();
+                let timeout = TimeoutType::IdleInTransactionSession(txn_id);
+                self.timeouts.add_timeout(timeout, timeout_dur);
+            }
+        }
+    }
+
+    pub fn remove_idle_in_transaction_session_timeout(&mut self) {
+        let session = self.session();
+        if let Some(txn) = session.transaction().inner() {
+            let txn_id = txn.id.clone();
+            self.timeouts
+                .remove_timeout(&TimeoutType::IdleInTransactionSession(txn_id));
+        }
+    }
+
+    pub async fn recv_timeout(&mut self) -> Option<TimeoutType> {
+        self.timeouts.recv().await
+    }
+
+    pub fn handle_timeout(&mut self, timeout: TimeoutType) -> Result<(), AdapterError> {
+        if self.timeouts.remove_timeout(&timeout) {
```

Comment:
This is an awkward API. Under what circumstances can a timeout be passed here that isn't in the hashmap?

Reply:
I think this and the comment below are related. If `timeout` is for a previous transaction, or the timeout has already been removed, then the `timeout` passed won't be in the hashmap. That's also why we don't check the txn id below, it's already checked in this remove call.

Though maybe we should add an assert that asserts the txn id matches the passed timeout.

Reply:
I updated this to remove this function. Now remove abort the task AND will drain the timeout queue of any equal timeouts. So there's no longer a race condition where we might receive a stale/old timeout.